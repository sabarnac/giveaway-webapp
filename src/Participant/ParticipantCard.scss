.participant-card {
  display: flex;
  flex-direction: column;
  justify-content: center;
  .participant-card__avatar {
    text-align: center;
    height: 5rem;
    img {
      height: 5rem;
    }
  }
  .participant-card__name {
    text-align: center;
    font-size: 3rem;
    word-break: break-all;
  }
  padding: 2rem 2rem 1rem 2rem;
  border: 1px solid #aaa;
  width: 45%;
}
