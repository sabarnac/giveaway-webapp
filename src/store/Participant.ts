import { observable, computed } from "mobx";

/**
 * Class containing the details of a participants' avatar.
 */
export class Avatar {
  @observable private _url: string;
  @observable private _altText: string;

  public constructor(url: string, altText: string) {
    this._url = url;
    this._altText = altText;
  }

  /**
   * Gets the avatar URL.
   * @return {string} The avatar URL.
   */
  public get url(): string {
    return this._url;
  }

  /**
   * Gets the avatar image alternate text.
   * @return {string} The avatar image alternate text.
   */
  public get altText(): string {
    return this._altText;
  }

  private isSameUrl = (url1: string, url2: string): boolean => url1 === url2;

  private isSameAltText = (altText1: string, altText2: string): boolean =>
    altText1 === altText2;

  /**
   * Determines whether another avatar is identical to the current one.
   * @param  {Avatar} otherAvatar The other avatar to compare against.
   * @returns {boolean} Whether the other avatar is equal to the current one.
   */
  public equals = (otherAvatar: Avatar): boolean =>
    this.isSameUrl(this._url, otherAvatar._url) &&
    this.isSameAltText(this._altText, otherAvatar._altText);
}

/**
 * Class containing the details of a participant.
 */
export default class Participant {
  @observable private _name: string;
  @observable private _avatar?: Avatar;

  public constructor(name: string, avatar?: Avatar) {
    this._name = name;
    this._avatar = avatar;
  }

  /**
   * Gets the participant name.
   * @return {string} The participant name.
   */
  public get name(): string {
    return this._name;
  }

  /**
   * Gets the participant avatar, if it exists.
   * @return {string | null} The participant avatar, or null if there isn't one.
   */
  public get avatar(): Avatar | null {
    return this._avatar || null;
  }

  /**
   * Gets whether the participant has an avatar or not.
   * @returns {boolean} Whether the participant has an avatar or not.
   */
  @computed public get hasAvatar(): boolean {
    return !!this._avatar;
  }

  /**
   * Gets the capitalized name of the participant.
   * @returns {string} The capitalized name.
   */
  @computed public get properName(): string {
    return this._name.replace(/\w\S*/g, (subText: string) => {
      return subText.charAt(0).toUpperCase() + subText.substr(1).toLowerCase();
    });
  }

  private isSameName = (name1: string, name2: string): boolean =>
    name1 === name2;

  private isSameAvatar = (avatar1?: Avatar, avatar2?: Avatar): boolean =>
    avatar1 && avatar2
      ? avatar1.equals(avatar2)
      : !avatar1 && !avatar2
      ? true
      : false;

  /**
   * Determines whether another participant is identical to the current one.
   * @param  {Participant} otherParticipant The other participant to compare against.
   * @returns {boolean} Whether the other participant is equal to the current one.
   */
  public equals = (otherParticipant: Participant): boolean =>
    this.isSameName(this._name, otherParticipant._name) &&
    this.isSameAvatar(this._avatar, otherParticipant._avatar);
}
