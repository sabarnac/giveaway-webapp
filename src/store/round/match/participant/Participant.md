Usage:

```tsx
import Avatar from "./store/round/match/participant/Avatar";

const participant: Participant = new Participant(
  "FOO BAR" /*, {url:"url",altText:"alt-text"} */
);
const participantName: string = participant.name; // "FOO BAR"
const participantProperName: string = participant.properName; // "Foo Bar"
const participantAvatar: Avatar = participant.avatar; // Avatar("url","alt-text")
```
