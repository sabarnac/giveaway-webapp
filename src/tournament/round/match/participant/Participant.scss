.participant {
    background: #ffffff;
    &--invert {
        background: #000000;
        color: #ffffff;
    }
}