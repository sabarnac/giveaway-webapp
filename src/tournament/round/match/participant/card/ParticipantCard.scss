@import '../Participant';
.participant-card {
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 2rem 2rem 1rem 2rem;
    width: 100%;
    border: 1px solid #000000;
    &--invert {
        border: 1px solid #ffffff;
    }
    &__avatar {
        text-align: center;
        height: 5rem;
        img {
            height: 5rem;
        }
    }
    &__name {
        text-align: center;
        font-size: 3rem;
        word-break: break-all;
    }
}