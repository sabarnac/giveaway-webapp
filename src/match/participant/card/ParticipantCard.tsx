import React, { Component } from "react";
import { observer } from "mobx-react";
import "./ParticipantCard.scss";
import classNames from "classnames";
import Participant from "../../../store/round/match/participant/Participant";

interface ParticipantCardProps {
  participant: Participant;
}

@observer
class ParticipantCard extends Component<ParticipantCardProps> {
  private getNameView = (): JSX.Element => (
    <div className={classNames("participant-card__name")}>
      <strong>{this.props.participant.name}</strong>
    </div>
  );

  private getAvatarImage = (): JSX.Element => (
    <img
      src={this.props.participant.avatar.url}
      alt={this.props.participant.avatar.altText}
    />
  );

  private getAvatarView = (): JSX.Element => (
    <div className={classNames("participant-card__avatar")}>
      {this.getAvatarImage()}
    </div>
  );

  public render = (): JSX.Element => {
    return (
      <div className={classNames("participant-card")}>
        {this.getAvatarView()}
        {this.getNameView()}
      </div>
    );
  };
}

export default ParticipantCard;
