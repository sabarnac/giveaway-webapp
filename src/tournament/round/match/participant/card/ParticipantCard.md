Usage:

```tsx
import Participant from "./store/round/match/participant/Participant";

const participant: Participant = new Participant("FOO");
<ParticipantCard participant={participant} inverse={false} />;
```
