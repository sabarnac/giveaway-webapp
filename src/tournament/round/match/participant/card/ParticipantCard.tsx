import React, { Component } from "react";
import { observer } from "mobx-react";
import "./ParticipantCard.scss";
import classNames from "classnames";
import Participant from "../../../../../store/round/match/participant/Participant";

/**
 * Properties of the participant card React component.
 */
interface ParticipantCardProps {
  /** The participant details. */
  participant: Participant;
  /** Whether to invert the color of the participant view or not. */
  invert?: boolean;
}

/**
 * React component for the participant card.
 */
@observer
export default class ParticipantCard extends Component<ParticipantCardProps> {
  /**
   * Returns the name view of the participant.
   * @return {JSX.Element} The name view.
   */
  private _getNameView = (): JSX.Element => (
    <div className={classNames("participant-card__name")}>
      <strong>{this.props.participant.properName}</strong>
    </div>
  );

  /**
   * Returns the avatar image view of the participant.
   * @return {JSX.Element} The avatar image view.
   */
  private _getAvatarImage = (): JSX.Element => (
    <img
      src={this.props.participant.avatar.url}
      alt={this.props.participant.avatar.altText}
    />
  );

  /**
   * Returns the avatar view of the participant.
   * @return {JSX.Element} The avatar view.
   */
  private _getAvatarView = (): JSX.Element => (
    <div className={classNames("participant-card__avatar")}>
      {this._getAvatarImage()}
    </div>
  );

  /**
   * Renders the component.
   * @return {JSX.Element} The rendered component.
   */
  public render = (): JSX.Element => (
    <div
      className={classNames("participant", "participant-card", {
        "participant--invert": this.props.invert,
        "participant-card--invert": this.props.invert,
      })}
    >
      {this._getAvatarView()}
      {this._getNameView()}
    </div>
  );
}
