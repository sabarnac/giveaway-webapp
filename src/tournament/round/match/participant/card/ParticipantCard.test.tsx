import React from "react";
import ParticipantCard from "./ParticipantCard";
import { create, ReactTestRenderer } from "react-test-renderer";
import { createDummyParticipant } from "../../../../../util/test/index";
import Participant from "../../../../../store/round/match/participant/Participant";

describe("ParticipantCard Component.", () => {
  it("Matches snapshot.", () => {
    const participant: Participant = createDummyParticipant();
    const component: ReactTestRenderer = create(
      <ParticipantCard participant={participant} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
