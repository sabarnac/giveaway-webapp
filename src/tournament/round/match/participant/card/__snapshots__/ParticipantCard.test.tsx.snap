// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParticipantCard Component. Matches snapshot. 1`] = `
<div
  className="participant participant-card"
>
  <div
    className="participant-card__avatar"
  >
    <img
      alt="foobar-1"
      src="foobar-1"
    />
  </div>
  <div
    className="participant-card__name"
  >
    <strong>
      Fozbaz 1
    </strong>
  </div>
</div>
`;
